### DISCUSSION TITLE
Set the "Column default value settings" using PnP Core SDK

### DISCUSSION THREAD

**Discussion by mvcsharepointdev** [pnp/pnpcore#1101] [Q&A]
We have a document library which have 2 managed metadata columns as follow:-

![image](https://user-images.githubusercontent.com/79726511/219800276-c6cbd5cc-c9e3-4773-85fa-bba89895bfad.png)

so can we set the default values for the folders using PnP Core SDK ? currently we do so from the UI using this link:-

![image](https://user-images.githubusercontent.com/79726511/219800463-9ee4bc0d-ff1f-4ddd-87eb-08d951c915c1.png)

any advice on this please?

**Comment by MathijsVerbeeck** [comment]
This is possible. Please check the following link: [https://pnp.github.io/pnpcore/using-the-sdk/lists-intro.html#configuring-default-values-for-a-list](https://pnp.github.io/pnpcore/using-the-sdk/lists-intro.html#configuring-default-values-for-a-list)

**Reply by mvcsharepointdev** [reply]
@MathijsVerbeeck thanks a lot for the link, and can we populate those default values using managed metadata columns? what we need to pass the term id for example?

**Reply by MathijsVerbeeck** [reply]
I am not sure about that, you would have to do some research for yourself about that how it's done in the backend.

**Reply by jansenbe** [reply]
Thanks for the providing the link @MathijsVerbeeck . @mvcsharepointdev : Using managed metadata is possible as well, the text format is slightly more complicated though. Here's the code from PnP Framework for this, notice the highlighted format. Please report back here once it's works, we can then update the docs to explain this.

![image](https://user-images.githubusercontent.com/7451219/220269470-b3d83354-6288-4d53-baa2-a270377a5021.png)

**Reply by mvcsharepointdev** [reply]
@jansenbe  here what i got  : - 

I am creating a new folder >> pass its path + field internalname + term id + the list instance to a method which will set the default values for the folder:-
```

var folder = (await context.Web.Lists.GetByTitleAsync(documentlibrarytitle, p => p.RootFolder)).RootFolder;

                                // Add the main folder represnting the customer name
                                var customer = (FieldTaxonomyValue)listItem["Customer"];
                                IFolder subFolder = await folder.Folders.AddAsync(customer.Label);
                               await SetDefaultAsync("/"+ subFolder.Name, "Customer",customer.TermId,myList);
```

and here is the SetDefaultAsync method:-
```

private async Task SetDefaultAsync(string folderpath, string fieldname, Guid defaultvalue, PnP.Core.Model.SharePoint.IList mylist)
        {
            List<DefaultColumnValueOptions> defaultColumnValues = new()
{
    new DefaultColumnValueOptions
    {
        FolderRelativePath = folderpath,
        FieldInternalName = fieldname,
        DefaultValue = defaultvalue.ToString()
    }
            };

            await mylist.SetDefaultColumnValuesAsync(defaultColumnValues);

        }
```

but i keep getting this error on ` await mylist.SetDefaultColumnValuesAsync(defaultColumnValues);`:-

```
{HttpResponseCode: 404
Code: System.IO.FileNotFoundException
Message: File Not Found.
ClientRequestId: 595999a0-8052-3000-5478-9e8765295aab
SPClientServiceRequestDuration: 21
X-SharePointHealthScore: 3
X-SP-SERVERSTATE: ReadOnly=0
}


```

I also tried to pass the value as follow:-

`await SetDefaultAsync("/"+ subFolder.Name, "Customer","-1;#"+customer.Label+"|"+customer.TermId.ToString()+";#",myList);`

but got the same error..
any advice please?
Thanks

**Reply by mvcsharepointdev** [reply]
@jansenbe appreciate if you can check my above comment.. thanks

**Reply by jansenbe** [reply]
@mvcsharepointdev : Don't you need to pass in the field name (e.g. "MetaData1") instead of "Customer"?

**Reply by mvcsharepointdev** [reply]
@jansenbe  Customer is the internal field name which i want to set its default value.. the original screen shot was just for clarification , this what i need to set:-

![image](https://user-images.githubusercontent.com/79726511/221249835-5427f4d8-0966-4f25-8b64-425cacb8c697.png)

&

![image](https://user-images.githubusercontent.com/79726511/221250338-e41689eb-f9b4-4f8a-a6e1-800da7700209.png)


Appreciate your advice.Thanks

**Reply by jansenbe** [reply]
Can you first try with a basic text field and see if you get that to work based upon our docs

**Reply by mvcsharepointdev** [reply]
@jansenbe ok sure will try it now

**Reply by mvcsharepointdev** [reply]
@jansenbe  tried it for a single line of text and i got the same error


{HttpResponseCode: 404
Code: System.IO.FileNotFoundException
Message: File Not Found.
ClientRequestId: 595999a0-8052-3000-5478-9e8765295aab
SPClientServiceRequestDuration: 21
X-SharePointHealthScore: 3
X-SP-SERVERSTATE: ReadOnly=0
}



what do you think the problem? Thanks

**Reply by mvcsharepointdev** [reply]
> This is possible. Please check the following link: https://pnp.github.io/pnpcore/using-the-sdk/lists-intro.html#configuring-default-values-for-a-list

Based on my test this is not possible . i test it using Managed Metadata field + Single Line of text field, as shown in the previous replies, are you able to advice plase? @MathijsVerbeeck

**Reply by jansenbe** [reply]
@mvcsharepointdev : I just ran our test again (https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/ListTests.cs#L1993-L2066) and that works great. Looking at your error I suspect this line to fail:

```csharp
// Get the folder to add the defaults file to
var formsFolder = await PnPContext.Web.GetFolderByServerRelativeUrlAsync($"{listInfo.RootFolder.ServerRelativeUrl}/Forms", p => p.Files).ConfigureAwait(false);
```

Are you using this on a document library or on a regular list? We don't have a "Forms" folder by default on a regular list and column defaults do not work there (also not from the UI). Can you cross check your list?

**Reply by mvcsharepointdev** [reply]
@jansenbe thanks for your reply.. so what i need to modify exactly inside my code? , this what i currently have:-

```
var folder = (await context.Web.Lists.GetByTitleAsync(documentlibrarytitle, p => p.RootFolder)).RootFolder;

                                // Add the main folder represnting the customer name
                                var customer = (FieldTaxonomyValue)listItem["Customer"];
                                IFolder subFolder = await folder.Folders.AddAsync(customer.Label);
                               await SetDefaultAsync("/"+ subFolder.Name, "Customer",customer.TermId,myList);
```


```
private async Task SetDefaultAsync(string folderpath, string fieldname, Guid defaultvalue, PnP.Core.Model.SharePoint.IList mylist)
        {
            List<DefaultColumnValueOptions> defaultColumnValues = new()
{
    new DefaultColumnValueOptions
    {
        FolderRelativePath = folderpath,
        FieldInternalName = fieldname,
        DefaultValue = defaultvalue.ToString()
    }
            };

            await mylist.SetDefaultColumnValuesAsync(defaultColumnValues);

        }
```

where i am passing the term id..

also i tried to pass this as a combination between termLabel + TermID+ -1:-
`
await SetDefaultAsync("/"+ subFolder.Name, "Customer","-1;#"+customer.Label+"|"+customer.TermId.ToString()+";#",myList);`

Thanks

**Reply by jansenbe** [reply]
Do you special chars or spaces in your folder names?

**Reply by mvcsharepointdev** [reply]
@jansenbe yes some folders have white spaces such as "customer A".. why ? is this a problem?

**Reply by mvcsharepointdev** [reply]
@jansenbe also where i need to insert this line of code inside my code:-

`var formsFolder = await PnPContext.Web.GetFolderByServerRelativeUrlAsync($"{listInfo.RootFolder.ServerRelativeUrl}/Forms", p => p.Files).ConfigureAwait(false);`

i am totally confused, is there any documentations i can refer to on how to set the default values for folders based on managed metadata columns, this will be very helpful? Thanks in advance for any help.

**Reply by mvcsharepointdev** [reply]
> @mvcsharepointdev : I just ran our test again (https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/ListTests.cs#L1993-L2066) and that works great. Looking at your error I suspect this line to fail:
> 
> ```cs
> // Get the folder to add the defaults file to
> var formsFolder = await PnPContext.Web.GetFolderByServerRelativeUrlAsync($"{listInfo.RootFolder.ServerRelativeUrl}/Forms", p => p.Files).ConfigureAwait(false);
> ```
> 
> Are you using this on a document library or on a regular list? We don't have a "Forms" folder by default on a regular list and column defaults do not work there (also not from the UI). Can you cross check your list?

i am using document library not a custom list..

also inside your test, seems you are adding a single line of text not a term..

**Reply by jansenbe** [reply]
@mvcsharepointdev : I've just updated our default value test case to also use folder names with spaces and a taxonomy field and all works fine. Here's the current test case: https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core.Test/SharePoint/ListTests.cs#L1994-L2101

Here's the default values set using that test:
![image](https://user-images.githubusercontent.com/7451219/221535671-ef5458f3-eba0-4a3f-9950-de1369b29b94.png)

I've updated the docs to explain the needed format for setting taxonomy field defaults: https://pnp.github.io/pnpcore/using-the-sdk/lists-intro.html#configuring-default-values-for-a-list

If you still can't get it to work then do a Fiddler trace or debug PnP Core SDK (see https://pnp.github.io/pnpcore/using-the-sdk/basics-debug.html) and provide the extra details.

**Reply by mvcsharepointdev** [reply]
@jansenbe working well now,, thanks a lot for your great help.

**Reply by mvcsharepointdev** [reply]
@jansenbe  i have noted something wired, is that when i set the default value for a folder it will remove all the previous default values from all the folder !! any advice?

**Reply by mvcsharepointdev** [reply]
@jansenbe  in CSOM i use to update the defualt values for the folders without causing the prevouse defualt to rest as follow:-
```

List listtooperate = context.Web.GetList(webrelativeurl + "/" + template["TemplateListUrl"].ToString());
                                MetadataDefaults bb = new MetadataDefaults(context, listtooperate);

                                if (listItem["CorpType"] != null)
                                {

                                    var corpTypetaxValue = listItem["CorpType"] as TaxonomyFieldValue;
                                    if (corpTypetaxValue != null)
                                    {
                                 
                                        bb.SetFieldDefault(context.Web.GetFolderByServerRelativeUrl(webrelativeurl + "/" + template["TemplateListUrl"].ToString() + "/" + mainfoldername.ToString()), "CorpType", "1033;#" + corpTypetaxValue.Label + "|" + corpTypetaxValue.TermGuid);
                                        bb.Update();

                                    }
                                }
```

so can i do the same using PnP core SDK?
Thanks

**Reply by jansenbe** [reply]
@mvcsharepointdev : see my comment on the issue you created for this

**Reply by mvcsharepointdev** [reply]
@jansenbe thanks, replied to your comment. please check it . thanks
